### DISCUSSION TITLE
How to calculate BPM from clock timecodes

### DISCUSSION THREAD

**Discussion by davay42** [djipco/webmidi#177] [Q&A]
Every synth with midi sync can easily establish the current BPM from the clock messages stream, but all my attempts to calculate it from clock timestamp events give a very unstable number and seem too heavy for the CPU. 

My current idea is to circularly fill an array with differences between current and previous clock timestamps and then iterate over that array to calculate the exact BPM. I came up to this formula `bpm = 60000 / ((ev.timestamp - prevTimestamp) * PPQ)` Where I got `PPQ = 24` for my OP-Z. But all my actual implementations are buggy and slow.

Is there an optimized way to calculate it more fast and precisely? I think it's a rather common question and there should be a community approved optimal way to get the number. May be some ArrayBuffer can make the calculations fast and reliable?

@djipco Do you have any experience with that?

**Comment by djipco** [comment]
> @djipco Do you have any experience with that?

I'm afraid I do not but it would be an interesting addition to WebMidi.js (in fact, I added it to the [Enhancements to evaluate](https://github.com/djipco/webmidi/wiki#enhancements-to-evaluate) section of the wiki).

If I had to do it, I would pretty much go the same route. That is:

1. Calculate the delta between clock messages using the event's `timestamp` property (not JavaScript's time!).
2. Save the delta in an array of the last deltas. The length of that array would need to be tested for. A longer array means better smoothing but also means slower response to change.
3. Calculate the average of the deltas
4. Your bpm would be: ` (1000 / averageClockDelta / 24) * 60`

Instead of using an array, you could also use a smoothing factor (which would probably be a cleaner solution). There is an example of this here:

https://github.com/yderidde/PGMidi/blob/master/Sources/PGMidi/PGMidiSession.mm#L186

If you ever end up with a satisfying solution and don't mind sharing, I would love to see what you came up with.

Cheers!

**Reply by davay42** [reply]
Thank you for the ideas and links to explore. The first thing I get from there is that the millisecond resolution isn't fine enough to get precise BPM. We need nanoseconds grid to get the right resolution. Any ideas on that? May be we should use  `window.performance.now()` for that?

**Reply by djipco** [reply]
The `timestamp` property is, indeed, expressed in milliseconds but it is a decimal number. As per the [Web MIDI API specification](https://webaudio.github.io/web-midi-api/#MIDIMessageEvent), the `timestamp` is a [DOMHighResTimeStamp](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp) whose precision should be accurate to 5 µs (microseconds). So, theoretically, it should be enough to get an accurate BPM measurement.

By the way, `WebMidi.time` is simply an alias to `performance.now()`.
